Support for OAuth2 Protected Resource Metadata (RFC 9728)

sberyozkin opened the pull request:
Closes: #47317.

This PR provides an initial support for [RFC9728](https://datatracker.ietf.org/doc/rfc9728/).

[RFC9728](https://datatracker.ietf.org/doc/rfc9728/) metadata document may include a lot of properties but this PR starts with returning an `authorization_servers` property only, to support the [MCP authorization requirement](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#overview) and specifically, the [requirement to support authorization_servers ](https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#authorization-server-location), to allow MCP clients bootstrap themselves with discovering how they can login users into MCP servers.

Going forward, more metadata may be included, given the specific requirements.

Right now, returning the resource metadata is disabled by default since this information can be considered sensitive. For example, `www.someservice.com` may not want to advertise on the web which authorization providers or endpoints and other related  metadata are supporting the user's logins... For example, with Keycloak or other production quality providers, by registering custom login pages, it can be hard to figure out what the actual provider is used to login users and therefore, it should be a deployment level decision whether to expose the protected resource metadata or not.
That might need to be relaxed though in some cases, may be in devmode.

`quarkus.oidc.resource-metadata.resource` property is optional.
If it is not configured then the handler path is calculated as follows: if it is a default tenant then it is just `/.well-known/oauth-protected-resource`, otherwise `/.well-known/oauth-protected-resource/${tenant-id}`.
If it is configured, then if it is an absolute URL then  it is either `/.well-known/oauth-protected-resource`, or `/.well-known/oauth-protected-resource/${url-path-component}`. If it is a relative URL, then it is `/.well-known/oauth-protected-resource`, otherwise `/.well-known/oauth-protected-resource/relative-url`.

I've added a few tests for static and dynamic tenants, and also confirmed with NGrok that configuring an absolute HTTPS based resource URL also works.

Right now I'm finalizing the spec option where the `WWW_Authenticate Bearer resource_metadata=...` is returned if no token is available when accessing the protected resource, which is also required by the MCP authorization; PR will be ready for review once it is done

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/ResourceMetadataHandler.java:
```diff
@@ -0,0 +1,164 @@
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+public class ResourceMetadataHandler implements Handler<RoutingContext> {
```
Yeah, it is not complete yet, was only starting with structuring it similar to the latest backchannel logout code

michalvavrik replied:
Got it.

sberyozkin commented:
@michalvavrik @pedroigor @gastaldi, Hi, it should help us get started with wiring in MCP clients like Claude with Quarkus MCP servers, once Claude is ready. As I said in the description, the actual metadata is limited to the list of authorization servers, since that is the only practical requirement I'm aware of at the moment, we can add more metadata as needed going forward...
Have a look when you get a chance

sberyozkin commented:
Thanks @gastaldi, I'll wait till tomorrow to see if more reviews are provided. I followed the pattern @michalvavrik applied in his PR to support dynamic back channel logouts, so if Michal could double check it would be good

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/ResourceMetadataHandler.java:
```diff
@@ -0,0 +1,242 @@
+        String rootPath = OidcCommonUtils.prependSlash(configuredRootPath);
+        // Strip trailing '/' if the length is > 1
+        if (rootPath.length() > 1 && rootPath.endsWith("/")) {
+            rootPath = rootPath.substring(rootPath.length() - 1);
```
@gastaldi I'll add a test, thanks for catching it

sberyozkin commented:
@gastaldi @michalvavrik, I cleaned it up, added a test for the root path calculation bug related to the trailing `/` (it came via the copy and paste of the BackChannel logout code that I added awhile back - luckily, cases like `/root/` have never happened - I'll create another PR to clean it up for the backchannel logout for 3.20/3.24)

sberyozkin committed:
Support for OAuth2 ResourceProtectedMetadata

michalvavrik commented:
thanks @sberyozkin , I'll review tomorrow

michalvavrik reviewed (approved):
LGTM, let's merge after you looked at my comments on minor thing.

sberyozkin commented:
@michalvavrik Thanks, I fixed the typos, removed the `public` qualifier, and we can look at supporting signed metadata when we can see more clarify about this option

sberyozkin commented on integration-tests/oidc-wiremock/src/main/java/io/quarkus/it/keycloak/CustomTenantConfigResolver.java:
```diff
@@ -107,6 +108,7 @@ public Uni<OidcTenantConfig> resolve(RoutingContext context,
             config.getCodeGrant().setExtraParams(Map.of("extra-param", "extra-param-value"));
             config.getAuthentication().setInternalIdTokenLifespan(Duration.ofSeconds(301));
             config.setAllowUserInfoCache(false);
+            config.resourceMetadata.enabled = true;
```
@michalvavrik No, you are totally right, we don't really want to continue promoting legacy approach, I probably add `public` by some copy and paste, let me handle it

michalvavrik replied:
Cool

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcTenantConfig.java:
```diff
@@ -141,6 +141,35 @@ interface IntrospectionCredentials {
+         * This property is ignored if {@link #resource() is an absolute URL}
+         */
+        @WithDefault("true")
+        boolean forceHttpsScheme();
```
@michalvavrik Sure, this is a very new spec, and MCP clients like Claude are likely to be the only visible testers/users of this spec for a while. We can indeed add a few more of those properties going forward

sberyozkin commented:
@jacobhm98 Hi, I missed your ping here but I believed we had the conversation in Discussions. Protected Resource Metadata is now supported, please open a bug issue if it does not work as expected
